### DISCUSSION TITLE
How to change the color of the dialog？

### DISCUSSION THREAD

**Discussion by 600888** [magiblot/tvision#107] [Q&A] [answered]
How to change the color of the dialog box, I have overridden the getPalette() method as required, but this does not work, the dialog box displays red by default, why?
![3](https://github.com/magiblot/tvision/assets/59718553/230bb4a9-748e-4b85-99d3-e9a9677aef0e)
![4](https://github.com/magiblot/tvision/assets/59718553/54e8a65e-6bbf-488b-9d7a-c216034cc646)

**Comment by magiblot** [comment]
Hi @600888!

Sorry for the late answer.

I am uncertain why you are facing this issue. The white-over-red color is automatically used when there is something wrong with the palette. However, I don't see anything wrong in your definition of `YcViewDialog::getPalette()`, so the problem is likely somewhere else. Is the `getPalette()` method overriden in any other classes? Which class are the `YcViewDialog` instances being inserted into?

Note, however, that if your only purpose is to use the blue dialog palette, you can also achieve this not by overriding `getPalette()`, but by setting the `palette` member variable ([inherited from `TWindow`](https://github.com/magiblot/tvision/blob/92177cb365b523b0bc3e17d865292045ed7e0073/include/tvision/views.h#L1006)) to the constant [`dpBlueDialog`](https://github.com/magiblot/tvision/blob/92177cb365b523b0bc3e17d865292045ed7e0073/include/tvision/dialogs.h#L95), like this:

```c++
class MyDialog : public TDialog
{
public:
    MyDialog(const TRect &bounds, TStringView title) :
        TWindowInit(initFrame),
        TDialog(bounds, title)
    {
        palette = dpBlueDialog;
    }
};
```

Cheers.

**Comment by 600888** [comment]
Thank you for your reply, but it seems that it didn't work. Let me describe my design in detail. Here is my design structure:

![页-1](https://github.com/magiblot/tvision/assets/59718553/2cfb2001-c5ee-4ef3-8577-9bb30c526efb)
First, I created a TAPP，like this:

![2](https://github.com/magiblot/tvision/assets/59718553/6913c7d1-e6d3-4881-bf9c-881d1fdf0921)

then Pressing F2 on the TAPP interface will open the StationViewDialog interface, as shown in the figure below:

![7](https://github.com/magiblot/tvision/assets/59718553/333a94c4-3718-406c-aacb-b34f23a825ff)
![3](https://github.com/magiblot/tvision/assets/59718553/f9a5602f-c5b5-4af2-a1f9-9cd1e7127a5c)

In the StationViewDialog class, I placed the YcViewDialog. Clicking on the location indicated by the red arrow in the figure can open the YcViewDialog. I want the color style of YcViewDialog to be blue like StationViewDialog, but it becomes red instead.
![5](https://github.com/magiblot/tvision/assets/59718553/74174e1e-12c7-4c50-9cff-553b009ff5e2)

![6](https://github.com/magiblot/tvision/assets/59718553/4f15d824-27ee-485f-9ea8-3674686ec681)

![4](https://github.com/magiblot/tvision/assets/59718553/f27d9e9f-5cad-40b0-940e-e6556b499fc0)

**Comment by magiblot** [comment, marked as answer]
Okay, I see where the problem is.

Here are some screenshots from the [Turbo Vision's programming guide](https://archive.org/details/bitsavers_borlandTurrogrammingGuide1992_25707423) that will help you understand how palettes actually work.

![Screenshot_20230725_083601](https://github.com/magiblot/tvision/assets/20713561/b7113571-0eed-4fa6-8ec6-ccfa1a548ea2)
![Screenshot_20230725_083615](https://github.com/magiblot/tvision/assets/20713561/0a1ef091-89e8-4800-844a-cf17c272e488)
![Screenshot_20230725_083649](https://github.com/magiblot/tvision/assets/20713561/e5b2422e-3ad3-4f94-8d83-52b0884c2295)

Turbo Vision's palette system has the disadvantage that when you design a palette, you have to take into account the class into which your view will be inserted.

The cause of your issue is that `YcViewDialog`s and `YxViewDialog`s are being inserted into `StationViewDialog`s, but still use standard `TDialog` palettes which assume the owner view will be a `TApplication` or `TDesktop`.

There are two possible basic solutions that come to my mind:
* Insert `YcViewDialog`s and `YxViewDialog`s into the desktop and not into another dialog (which is the most common practice):
    ```c++
    TProgram::deskTop->insert(ycViewDialog);
    ```
    Or, if the dialog must be modal:
    ```c++
    TProgram::deskTop->execView(ycViewDialog);
    ```
* Modify `YcViewDialog`'s and `YxViewDialog`'s `getPalette()` method into this:
    ```c++
    virtual TPalette &getPalette() const override
    {
        return TView::getPalette();
    }
    ```
    With this you won't see the white-over-red color issue but you will keep the same colors as `StationViewDialog`.

Cheers.

**Reply by 600888** [reply]
Thank you for your patient help, the problem has been successfully solved now!
![1](https://github.com/magiblot/tvision/assets/59718553/9f75ccf6-c209-4170-904c-5d2593b5b8e7)
